### ISSUE TITLE
Integration with Meteor.js

### ISSUE THREAD

**Issue by splendido** [Semantic-Org/Semantic-UI#1298]
Hi all,

I'd like to propose to start an official action for the integration of this awesome _User Interface_ into the  [Meteor.js](meteor.com), the [most popular full-stack JavaScript framework](https://github.com/meteor/meteor).

I'm starting off based on [this very good attempt](https://github.com/percolatestudio/atmosphere-beta/issues/297), about trying to put some order into the [packages database](https://atmospherejs.com) for Meteor, started by @dandv.
He also promoted an official integration for FontAwesome opening [this issue](https://github.com/FortAwesome/Font-Awesome/pull/5062) on [their official GitHub repository](https://github.com/FortAwesome/Font-Awesome) which so far had nothing to do with Meteor.

Since [he wrote his proposal](https://github.com/FortAwesome/Font-Awesome/pull/5062#issue-50168664) in a very clean wording, I ask you to read it carefully since it can be exactly the same for Semantic-UI.

At the moment [there are at least five different packages](https://atmospherejs.com/?q=semantic-ui) all simply wrapping semantic source code.

@lumatijev, @nooitaf are the maintainer of the most stared packages: I hope they'll also agree about taking a single official action for this.

My proposal, if you agree, is as follows:
- verify whether the name 'semantic-ui' is still available for a new Meteor _organization_
- create the organization on meteor.com (it could be me if you wish, but not necessarily...)
- add @jlukic among the people belonging to the organization (requires him to create his own developer account) and possibly other people from both semantic-ui and meteor
- figure out the best pattern to follow the make it as easy as possible to publish the official wrapper package (possibly on [the lines suggested](https://github.com/FortAwesome/Font-Awesome/pull/5062) by @dandv) 
- publish the needed packages (css, less, sass, fonts, no fonts, etc...)
- contact all other developer who published similar packages and kindly ask to mark them as no more maintaned (so to have them disappearing from the list on Atmosphere)
- keep up the collaboration to maintain the published packages at best

Hope this make sense also to you...

Please let me know what do you think about this!

**Comment by PavanKu** [comment]
Guys, can anyone give some update about meteor integration? what is needed to make this fast?

**Comment by nooitaf** [comment]
@PavanKu Main Discussion is at https://github.com/Semantic-Org/Semantic-UI-Meteor/issues/1

@splendido did a good start on how to make publishing easy for the maintainer, but the 'themeing' issue is still unclear i think.

**Comment by dalgard** [comment]
The way I understand it, theming is mostly about overwriting variables in the LESS files of Semantic UI.

Does LESS have the same function as Sass does, where a variable can hold a default value, which may be overridden anywhere in the code, even before?

In that case, would it be a feasible way to go having a LESS file with variable overrides somewhere in app space and then compiling it before the LESS files in package space during the build?

I have no idea how or if the LESS package deals with uncompiled LESS code in packages (probably it doesn't), but maybe this should change in order to support our use case here.

**Comment by splendido** [comment]
We're waiting @jlukic to finish the [integration PR](https://github.com/Semantic-Org/Semantic-UI/pull/1607). He already started to merge it manually on the [meteor branch](https://github.com/Semantic-Org/Semantic-UI/tree/meteor) but I guess he's very busy these days with the 1.7.x releases going on...

After this, we should get official packages with plain CSS style files.


The problem with the LESS variable customization is better summarize in [this post](https://github.com/Semantic-Org/Semantic-UI-Meteor/issues/1#issuecomment-66493079) by Jack where it make clear the inheritance problems with LESS with the current structure of the project.

The good news might be that, once there will be a `pacakge.js` file on the master branch, if you download the Semantic bundle, customize it as you wish, and eventually build it, you can simply simlink the Semantic folder under your app's `packages` folder and your customization will be *imported* as a meteor package.

I'm more and more convinced that publishing different compiled themes or trying to find a way to let it easily customizable might be useless in view of this possibility.

Basically you could do something like this:

```shell
meteor create myApp
cd myApp
mkdir packages
cd packages
git clone https://github.com/Semantic-Org/Semantic-UI.git
cd Semantic-UI
....your customizations here ...
npm install
gulp install
cd ../../
meteor add semantic:ui
meteor
```

What do you think about this?

**Comment by serkandurusoy** [comment]
I spent the better half of this day trying to come up with a good strategy to include the official (as of yet) package and apply theming.

I certainly did want to avoid using gulp grunt bower etc and the likes. I love meteor for its simplicity and its isomorphic nature where I don't need to master different tools for different tasks and I love that it is itself perfectly capable of building various sources.

Meteor goes to great lengths to integrate everything into a single meteor command. No node install! No mongo install! No cordova tooling install! No android!... You see where I'm getting at.

If only we had means to `meteor add` themes, or perhaps a starter theme (css/less/sass) that we can copy over and begin hacking away...

You know what, I'm almost two years into meteor development, and I'm still dead-scared of node, npm, gulp, bower...

There... I said it...

;)

**Comment by jongold** [comment]
@splendido is there anything we can do whilst waiting for #1607 to be merged to get Semantic UI working in Meteor with variables & customization etc?

**Comment by jlukic** [comment]
You can always drop in a zip file.

**Comment by splendido** [comment]
Hi @jongold,
what do you mean with
> with variables & customization etc?
if you're thinking about a LESS version, feel free to investigate. Keep in mind that, from the previous discussions it seems hard to get a LESS version properly working because of a number of variables overwrites done on a component basis.

Unfortunately I'm not going to have time to look also into this: getting the canonical CSS packages (auto)published would be my *personal* goal... :(

Any news @jlukic on the merging process of `meteor-integration`?
Please let us know if there's something you're not comfortable with, of if you simply don't have enough time to look at this now.

**Comment by serkandurusoy** [comment]
Is there a suggested way of using one of the pre-compiled templates to use as a base for a custom theme?

Of course I'm asking for a way of doing it within meteor, not by cloning the semantic-ui repo.

Basically, what next after `meteor add nootifaf:semantic-ui`?

I don't quite understand what zip file to drop in where.

**Comment by dalgard** [comment]
In the project I'm working on, I'm simply overriding whatever parts of the Semantic UI basic theme I don't like with simple CSS declarations. It is working okay, even if it's a bit bothersome and doesn't produce the nicest code...

**Comment by serkandurusoy** [comment]
> a bit bothersome and doesn't produce the nicest code

nicely put :) Ok, I guess it at least seems to be the most *convenient* among the alternatives.

**Comment by splendido** [comment]
@nooitaf had a [semantic-ui-less](https://atmospherejs.com/nooitaf/semantic-ui-less) package (I never tried).
Perhaps someone played a bit with it? Was it doing its work? Were customization possible (to some extent)?

**Comment by nooitaf** [comment]
@serkandurusoy i would suggest not using a package but just use the src folder from the semantic-ui source and add the less package, so it gets compiled by meteor. I didn't try it yet, but it should work. 
https://github.com/nooitaf/meteor-semantic-ui/issues/27#issuecomment-64902611

The semantic-ui-less package is deprecated, because the 1.0 source works differently somehow.

**Comment by jlukic** [comment]
@splendido I'm trying to rewrite the gulpfile while doing the meteor integration. I want to move all the tasks to individual js files and use `requiteDir()`. I also need to update `gulp release` to create GitHub tags when deploying to each of the 50+ component libs. 

If everyone prefers I can just drop in a package.js file and make things available immediately. I have not personally done any testing with using SUI package in Meteor however.

I'd prefer to hold off until I can find a way to make LESS customization/build tools available in an easy to use set-up within a meteor package.

I'm trying to get `1.8.0` out the door today. You can see the [upcoming Release Notes](https://github.com/Semantic-Org/Semantic-UI/blob/next/RELEASE-NOTES.md#version-180---january-21-2015) to see what it covers. I need to spend the rest of the afternoon documenting some changes, doing some sanity checks, and other tedium.

**Comment by serkandurusoy** [comment]
@nooitaf thanks, and thanks for providing the semantic-ui package all this time as well.

But in light of @splendido 's comments:

> Keep in mind that, from the previous discussions it seems hard to get a LESS version properly working because of a number of variables overwrites done on a component basis.

I feer that path may lead to even more problems.

Also, I will lose the convenience of `meteor update`.

As it currently stands, I fear there are two unquestionable options:
* Use semantic-ui sources to build your theme as a standalone project and than include it in meteor
* Or use the package from meteor and stick to whatever it provides

Yes there is a vast middle ground there, but they are all kind of hacks, overwrites and non-standard ways of achieving the goal.

**Comment by serkandurusoy** [comment]
> I'm trying to get 1.8.0 out the door today.
@jlukic :+1:  for that.

The discussion around Meteor integration and my current frustrations do not overshadow my appreciation for what you've been doing both with the framework itself and your embracing response with third party integrations. Kudos for that!

**Comment by splendido** [comment]
I second @serkandurusoy in complimenting with Jack!

@jlukic, perhaps dropping in the `package.js` file into the root folder wouldn't be a bad idea?
...this way we can start publishing the first `semantic:ui@1.8.0` (by hands) and enabling the possibility to whom is interested in custumization to follow the small *guide* I posted in this [previous post](https://github.com/Semantic-Org/Semantic-UI/issues/1298#issuecomment-70243004) of mine.

So we can start picking up some feed-back on it.

What do you think?

**Comment by splendido** [comment]
Before getting able to publish a customizable LESS package I really think we'd need to waint until things get improoved a bit on this side.

There's a lot of active discussion about changing the way packages work, especially about letting the possibility to configure them before load time (in our case with SUI this would mean letting the possibility to specify some variables overwrite before LESS compilation...).

See for example [this card](https://trello.com/c/mHK2dpr5/68-new-way-of-defining-packages-and-controlling-file-load-order) on the official trello roadmap or [this discussion](https://github.com/MeteorCommunity/discussions/issues/11) and [this one](https://github.com/MeteorCommunity/discussions/issues/58) on [Meteor Community](https://github.com/MeteorCommunity)

In the meanwhile I suggest to take a look at how Nemo64 allows for customization in his [bootstrap package](https://github.com/Nemo64/meteor-bootstrap/)
As I already said, unfortunately, I'm not going to have enough time to explore this path, so please feel free to go ahead on this.

**Comment by jlukic** [comment]
I dropped `1.8.0`. Will weigh this all next week. I think all package managers are having the same issues. Build tools are locked inside PM folder which is treated akin to a `temp/` directory. 

I think the solve is the install process dropping the entire payload somewhere else. Preferably whatever it assumes is project root.

**Comment by splendido** [comment]
Congrats on the `1.8.0` release!

**Comment by splendido** [comment]
For the Meteor package, please consider that the `package.js` file is used only once when you run `meteor publish`:

at that time the list of added files is picked up from the package folder, bundled with [isobuild](https://www.meteor.com/isobuild) and uploaded to the [packages server](https://www.meteor.com/services/package-server).
Later on the `package.js` doesn't exists anymore.

Hence the only needs are:
1) have the compiled files under the `dist` directory correctly linked from inside the `package.js` file by means of relative paths
2) the compiled files exists at the moment of running `meteor publish`


So, to publish the SUI Meteor package, there are no build processes involved if you ensure to push compiled files to the repository everytime you make a new release.
Once [autopublish.meteor.com](http://autopublish.meteor.com/) will receive the notice, the repo will be cloned and `meteor publish` run from its root folder.
